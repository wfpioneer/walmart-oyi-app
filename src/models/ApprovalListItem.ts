export interface ApprovalListItem {
  imageUrl?: string;
  itemName: string;
  itemNbr: number;
  upcNbr: number;
  categoryNbr: number;
  categoryDescription: string;
  subCategoryNbr: number;
  subCategoryDescription: string;
  oldQuantity: number;
  newQuantity: number;
  dollarChange: number;
  initiatedUserId: string;
  initiatedTimestamp: string;
  approvalStatus: approvalStatus;
  approvalRequestSource: approvalRequestSource;
  isChecked?: boolean;
  daysLeft?: number;
}

export interface ApprovalCategory extends ApprovalListItem {
  categoryHeader?: boolean;
}

export enum approvalStatus {
  'Pending' = 'Pending',
  'Approved' = 'Approved',
  'Rejected' = 'Rejected',
  'Expired' = 'Expired'
}
export enum approvalRequestSource {
  'ItemDetails' = 'itemdetails',
  'Audits' = 'audits',

}
