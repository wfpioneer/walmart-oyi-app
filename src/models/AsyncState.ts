export interface AsyncState{
  isWaiting: boolean;
  value: any;
  error: any;
  result: any;
}
