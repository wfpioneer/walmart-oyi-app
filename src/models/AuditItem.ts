export interface ItemPalletInfo {
  palletId: string,
  quantity: number,
  sectionId: number,
  locationName: string,
  mixedPallet: boolean,
  newQty?: number
}
