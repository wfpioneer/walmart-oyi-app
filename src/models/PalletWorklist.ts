export type PalletWorklistType = 'MP';

export interface MissingPalletWorklistItemI {
  palletId: number;
  lastKnownPalletLocationId: number;
  lastKnownPalletLocationName: string;
  createUserId: string;
  createTs: string;
  palletDeleted: boolean;
  completed: boolean;
  completedTs?: string;
  completedUserId?: string;
  itemCount?: number;
  sectionID?: number;
}

// eslint-disable-next-line no-shadow
export enum Tabs {
  TODO='ToDo',
  COMPLETED='Completed'
}
