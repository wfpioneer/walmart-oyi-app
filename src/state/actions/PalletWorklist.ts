import { Tabs } from '../../models/PalletWorklist';

export const SET_SELECTED_TAB = 'PALLET_WORKLIST/SET_SELECTED_TAB';
export const SET_SELECTED_WORKLIST_PALLET_ID = 'PALLET_WORKLIST/SET_SELECTED_WORKLIST_PALLET_ID';
export const CLEAR_SELECTED_PALLET_WORKLIST_ID = 'PALLET_WORKLIST/CLEAR_SELECTED_PALLET_WORKLIST_ID';

export const setSelectedTab = (tab: Tabs) => ({
  type: SET_SELECTED_TAB,
  payload: tab
} as const);

export const setSelectedWorklistPalletId = (id: string) => ({
  type: SET_SELECTED_WORKLIST_PALLET_ID,
  payload: id
} as const);

export const clearSelectedWorklistPalletId = () => ({
  type: CLEAR_SELECTED_PALLET_WORKLIST_ID
} as const);

export type Actions =
  | ReturnType<typeof setSelectedTab>
  | ReturnType<typeof setSelectedWorklistPalletId>
  | ReturnType<typeof clearSelectedWorklistPalletId>
