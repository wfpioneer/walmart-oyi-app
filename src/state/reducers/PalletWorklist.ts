import { Tabs } from '../../models/PalletWorklist';
import {
  Actions,
  CLEAR_SELECTED_PALLET_WORKLIST_ID,
  SET_SELECTED_TAB,
  SET_SELECTED_WORKLIST_PALLET_ID
} from '../actions/PalletWorklist';

export interface PalletWorklistState {
   selectedTab: Tabs;
   selectedWorklistPalletId: string
  }

export const initialState: PalletWorklistState = {
  selectedTab: Tabs.TODO,
  selectedWorklistPalletId: ''
};

export const PalletWorklist = (state = initialState, action: Actions) => {
  switch (action.type) {
    case SET_SELECTED_WORKLIST_PALLET_ID:
      return {
        ...state,
        selectedWorklistPalletId: action.payload
      };
    case SET_SELECTED_TAB:
      return {
        ...state,
        selectedTab: action.payload
      };
    case CLEAR_SELECTED_PALLET_WORKLIST_ID:
      return {
        ...state,
        selectedWorklistPalletId: ''
      };
    default:
      return state;
  }
};
