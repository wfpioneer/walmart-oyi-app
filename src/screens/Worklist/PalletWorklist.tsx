import React, {
  Dispatch, EffectCallback, useEffect, useState
} from 'react';
import { FlatList, Text, View } from 'react-native';
import { useDispatch } from 'react-redux';
import Toast from 'react-native-toast-message';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import MissingPalletWorklistCard from '../../components/MissingPalletWorklistCard/MissingPalletWorklistCard';
import { strings } from '../../locales';
import { MissingPalletWorklistItemI } from '../../models/WorklistItem';
import { CustomModalComponent } from '../Modal/Modal';
import { styles } from './PalletWorklist.style';
import Button from '../../components/buttons/Button';
import COLOR from '../../themes/Color';
import { clearPallet } from '../../state/actions/saga';
import { useTypedSelector } from '../../state/reducers/RootReducer';
import { AsyncState } from '../../models/AsyncState';
import {
  hideActivityModal,
  showActivityModal
} from '../../state/actions/Modal';
import { CLEAR_PALLET } from '../../state/actions/asyncAPI';

interface PalletWorkListProps {
  palletWorklist: MissingPalletWorklistItemI[];
  displayConfirmation: boolean;
  setDisplayConfirmation: React.Dispatch<React.SetStateAction<boolean>>;
  dispatch: Dispatch<any>;
  clearPalletAPI: AsyncState;
  navigation: NavigationProp<any>;
  useEffectHook: (effect: EffectCallback, deps?: ReadonlyArray<any>) => void;
}

export const clearPalletAPIHook = (
  clearPalletApi: AsyncState,
  palletId: string,
  navigation: NavigationProp<any>,
  dispatch: Dispatch<any>,
  setDisplayConfirmation: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  if (navigation.isFocused()) {
    if (!clearPalletApi.isWaiting) {
      // Success
      if (clearPalletApi.result) {
        dispatch(hideActivityModal());
        setDisplayConfirmation(false);
        dispatch({ type: CLEAR_PALLET.RESET });

        // TODO REFRESH/CALL MISSING PALLET WORKLIST API

        Toast.show({
          type: 'success',
          text1: strings('PALLET.CLEAR_PALLET_SUCCESS', { palletId }),
          position: 'bottom'
        });
      }

      // Failure
      if (clearPalletApi.error) {
        dispatch(hideActivityModal());
        setDisplayConfirmation(false);
        dispatch({ type: CLEAR_PALLET.RESET });
        Toast.show({
          type: 'error',
          text1: strings('PALLET.CLEAR_PALLET_ERROR'),
          text2: strings('GENERICS.TRY_AGAIN'),
          position: 'bottom'
        });
      }
    } else {
      dispatch(showActivityModal());
    }
  }
};

export const PalletWorklistScreen = (props: PalletWorkListProps) => {
  const {
    clearPalletAPI,
    displayConfirmation,
    dispatch,
    palletWorklist,
    setDisplayConfirmation,
    navigation,
    useEffectHook
  } = props;
  let deletePalletId = '';

  useEffectHook(() => clearPalletAPIHook(
    clearPalletAPI,
    deletePalletId,
    navigation,
    dispatch,
    setDisplayConfirmation
  ),
  [clearPalletAPI]);

  const onDeletePress = () => {
    dispatch(clearPallet({ palletId: deletePalletId }));
  };

  const handleDeleteClick = (palletId: string) => {
    setDisplayConfirmation(true);
    deletePalletId = palletId;
  };

  const handleAddLocationClick = () => navigation.navigate('ScanPallet');

  return (
    <View style={styles.container}>
      <CustomModalComponent
        isVisible={displayConfirmation}
        modalType="FormHeader"
        onClose={() => {}}
      >
        <View style={styles.delHeader}>
          <Text style={styles.message}>
            {strings('MISSING_PALLET_WORKLIST.DELETE_PALLET')}
          </Text>
        </View>
        <Text style={styles.delText}>
          {strings('MISSING_PALLET_WORKLIST.DELETE_PALLET_CONFIRMATION')}
        </Text>
        <View style={styles.buttonContainer}>
          <Button
            style={styles.delButton}
            title={strings('GENERICS.CANCEL')}
            backgroundColor={COLOR.MAIN_THEME_COLOR}
            onPress={() => setDisplayConfirmation(false)}
          />
          <Button
            style={styles.delButton}
            title={strings('GENERICS.OK')}
            backgroundColor={COLOR.TRACKER_RED}
            onPress={() => onDeletePress()}
          />
        </View>
      </CustomModalComponent>
      <FlatList
        data={palletWorklist}
        keyExtractor={(item: MissingPalletWorklistItemI, index: number) => item.palletId + index.toString()}
        renderItem={({ item }) => (
          <MissingPalletWorklistCard
            palletId={item.palletId}
            lastLocation={item.lastKnownLocationName}
            reportedBy={item.createId}
            reportedDate={item.createTS}
            expanded={true} // TODO Toggle for a single Pallet WorkList Item
            addCallback={handleAddLocationClick}
            deleteCallback={() => handleDeleteClick(item.palletId.toString())}
            navigateCallback={() => {}}
          />
        )}
        onRefresh={null}
        refreshing={undefined}
        style={styles.list}
      />
    </View>
  );
};

export const PalletWorkList = () => {
  const [displayConfirmation, setDisplayConfirmation] = useState(false);
  const clearPalletAPI = useTypedSelector(state => state.async.clearPallet);
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const mockMPWorklist: MissingPalletWorklistItemI[] = [
    {
      createId: '11',
      createTS: '26/06/2022',
      lastKnownLocationId: 1,
      lastKnownLocationName: 'A1-1',
      palletDeleted: false,
      palletId: 7988,
      worklistType: 'MP',
      completed: undefined,
      completedId: undefined,
      completedTS: undefined
    }
  ];
  return (
    <PalletWorklistScreen
      palletWorklist={mockMPWorklist}
      displayConfirmation={displayConfirmation}
      setDisplayConfirmation={setDisplayConfirmation}
      dispatch={dispatch}
      clearPalletAPI={clearPalletAPI}
      navigation={navigation}
      useEffectHook={useEffect}
    />
  );
};
