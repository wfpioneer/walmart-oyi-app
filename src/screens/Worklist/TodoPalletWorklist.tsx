import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { Dispatch } from 'redux';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import { MissingPalletWorklistItemI } from '../../models/WorklistItem';
import { useTypedSelector } from '../../state/reducers/RootReducer';
import { PalletWorklist } from './PalletWorklist';
import { AsyncState } from '../../models/AsyncState';
import { mockMissingPalletWorklistTodo } from '../../mockData/mockWorkList';

interface TodoWorklistProps {
  getMPWorklistApi: AsyncState;
  displayConfirmation: boolean;
  setDisplayConfirmation: React.Dispatch<React.SetStateAction<boolean>>;
  dispatch: Dispatch<any>;
  clearPalletAPI: AsyncState;
  navigation: NavigationProp<any>;
}

export const TodoPalletWorklistScreen = (
  props: TodoWorklistProps
): JSX.Element => {
  const {
    clearPalletAPI,
    displayConfirmation,
    dispatch,
    getMPWorklistApi,
    setDisplayConfirmation,
    navigation
  } = props;

  let todoPalletData: MissingPalletWorklistItemI[] | undefined;
  // TODO remove when implementing getPalletWorklist items api call
  todoPalletData = mockMissingPalletWorklistTodo;

  if (getMPWorklistApi.result && getMPWorklistApi.result.data) {
    todoPalletData = getMPWorklistApi.result.data.filter(
      (item: MissingPalletWorklistItemI) => item.completed === false
    );
  }
  return (
    <PalletWorklist
      palletWorklist={todoPalletData}
      displayConfirmation={displayConfirmation}
      setDisplayConfirmation={setDisplayConfirmation}
      dispatch={dispatch}
      clearPalletAPI={clearPalletAPI}
      navigation={navigation}
      useEffectHook={useEffect}
    />
  );
};

export const TodoPalletWorklist = (): JSX.Element => {
  const getMPWorklistApi = useTypedSelector(state => state.async.getWorklist);
  const clearPalletAPI = useTypedSelector(state => state.async.clearPallet);
  const [displayConfirmation, setDisplayConfirmation] = useState(false);
  const dispatch = useDispatch();
  const navigation = useNavigation();
  return (
    <TodoPalletWorklistScreen
      dispatch={dispatch}
      clearPalletAPI={clearPalletAPI}
      displayConfirmation={displayConfirmation}
      setDisplayConfirmation={setDisplayConfirmation}
      getMPWorklistApi={getMPWorklistApi}
      navigation={navigation}
    />
  );
};
