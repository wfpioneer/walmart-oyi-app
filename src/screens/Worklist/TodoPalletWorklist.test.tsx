import React from 'react';
import { NavigationProp } from '@react-navigation/native';
import ShallowRenderer from 'react-test-renderer/shallow';
import { TodoPalletWorklistScreen } from './TodoPalletWorklist';
import { AsyncState } from '../../models/AsyncState';
import {
  mockMissingPalletWorklistComplete,
  mockMissingPalletWorklistTodo
} from '../../mockData/mockWorkList';
import { Tabs } from '../../models/PalletWorklist';

let navigationProp: NavigationProp<any>;

describe('TodoPalletWorklistScreen', () => {
  const defaultAsyncState: AsyncState = {
    isWaiting: false,
    error: null,
    value: null,
    result: null
  };
  const mockTrackEventCall = jest.fn();

  it('Renders an array of incomplete pallet worklist items with incomplete and complete items', () => {
    const renderer = ShallowRenderer.createRenderer();
    const getPalletWorklistSuccess: AsyncState = {
      ...defaultAsyncState,
      result: {
        data: [
          ...mockMissingPalletWorklistTodo,
          ...mockMissingPalletWorklistComplete
        ]
      }
    };
    renderer.render(
      <TodoPalletWorklistScreen
        dispatch={jest.fn()}
        clearPalletAPI={defaultAsyncState}
        displayConfirmation={false}
        setDisplayConfirmation={jest.fn()}
        getMPWorklistApi={getPalletWorklistSuccess}
        navigation={navigationProp}
        groupToggle={false}
        updateGroupToggle={jest.fn()}
        selectedTab={Tabs.COMPLETED}
        setPalletClicked={jest.fn()}
        trackEventCall={mockTrackEventCall}
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
