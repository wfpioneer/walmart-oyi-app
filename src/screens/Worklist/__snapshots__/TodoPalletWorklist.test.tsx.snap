// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPalletWorklistScreen Renders an array of incomplete pallet worklist items with incomplete and complete items 1`] = `
<PalletWorklist
  clearPalletAPI={
    Object {
      "error": null,
      "isWaiting": false,
      "result": null,
      "value": null,
    }
  }
  dispatch={[MockFunction]}
  displayConfirmation={false}
  error={null}
  onRefresh={[Function]}
  palletWorklist={
    Array [
      Object {
        "completed": false,
        "completedTs": undefined,
        "completedUserId": undefined,
        "createTs": "26/06/2022",
        "createUserId": "11",
        "lastKnownPalletLocationId": 1,
        "lastKnownPalletLocationName": "A1-1",
        "palletDeleted": false,
        "palletId": 7988,
        "worklistType": "MP",
      },
      Object {
        "completed": false,
        "completedTs": undefined,
        "completedUserId": undefined,
        "createTs": "26/06/2022",
        "createUserId": "12",
        "lastKnownPalletLocationId": 2,
        "lastKnownPalletLocationName": "A1-2",
        "palletDeleted": false,
        "palletId": 7989,
        "worklistType": "MP",
      },
    ]
  }
  refreshing={false}
  setDisplayConfirmation={[MockFunction]}
  useEffectHook={[Function]}
/>
`;
